activedata_etl/transforms/per_test_to_es.py

```diff
@@ -0,0 +1,220 @@
+            Log.warning("per-test-coverage {{url}} has uncoverable lines", url=artifact.url)
+
+        new_record = set_default(
+            {
+                "source": {
+                    "file": set_default(
+                        file_details,
+                        {
+                            "covered": covered,
+                            "uncovered": uncovered,
+                            "total_covered": len(covered),
+                            "total_uncovered": len(uncovered),
+                            "percentage_covered": len(covered) / coverable_lines if coverable_lines else None
+                        }
+                    )
+                },
+                "etl": {
+                    "id": count(),
+                    "source": parent_etl,
+                    "type": "join",
+                    "machine": machine_metadata,
+                    "timestamp": Date.now()
+                }
+            },
+            task_cluster_record
+        )
+
+        return new_record
+
+    def process_source_file(parent_etl, count, test, sf):
+        sf = wrap(sf)
+
+        covered = []
+        uncovered = []
+        coverable = []
+        for l, c in enumerate(sf["coverage"]):
+            if c is None:
+                continue
+            coverable.append(l)
+            if c > 0:
+                covered.append(l)
+            else:
+                uncovered.append(l)
+
+        # turn covered into a set for use later
+        file_covered = set(covered)
+        coverable = set(coverable)
+
+        record = create_record(parent_etl, count, sf["name"], covered, uncovered)
+        record.test = {
+            "name": test,
+        }
+
+        # orphan lines (i.e. lines without a method), initialized to all lines
+        orphan_covered = set(covered)
+        orphan_uncovered = set(uncovered)
+
+        # iterate through the methods of this source file
+        # a variable to count the number of lines so far for this source file
+        methods = sf['functions'] if 'functions' in sf else []
+        method_start_indexes = [method['start'] for method in methods]
+        end = len(sf["coverage"])
+        for method in methods:
+            func_start = method['start']
+            func_end = end
+
+            for start in method_start_indexes:
+                if start > func_start:
+                    func_end = start
+                    break
+
+            method_lines = []
+            for l in coverable:
+                if l < func_start:
+                    continue
+
+                if l >= func_end:
+                    break
+
+                method_lines.append(l)
+
+            all_method_lines = set(method_lines)
+            method_covered = all_method_lines & file_covered
+            method_uncovered = all_method_lines - method_covered
+            method_percentage_covered = len(method_covered) / len(all_method_lines)
```

Comment:
set `method_percentage_covered` to `None` when `len(all_method_lines)` is zero

Reply:
Done!